### PR TITLE
Defsyncgateway

### PR DESCRIPTION
**Pull request by paulrdouglas**
Added deferred sync capability.

Changed response format to align with `sync` input for `syncexec/asyncexec`. 

Keeps backward compatibility with v3.5

### CODE REVIEW

File: code/processes/gateway.q [hunk 1]
```diff
@@ -131,7 +130,7 @@ canberun:{
  // check if it's possible to run anything
  availServers:availableserverstable[1b];
  if[0=count avail:distinct exec servertype from availServers;
-	:update required:(),available:(),handles:() from 0#0!queryqueue];
+  :update required:(),available:(),handles:() from 0#0!queryqueue];
   availIDs:exec serverid from availServers;
```
Comment (ThomasSmyth) [inline code/processes/gateway.q]: New line for closing square to make it more obvious where this condition ends.

File: code/processes/gateway.q [hunk 2]
```diff
@@ -398,73 +402,79 @@ getserveridstype:{[att;typ]
...
-		exec serverid!attributes from .gw.servers where active;
-		exec serverid!attributes from .gw.servers where active, servertype=typ];
+    exec serverid!attributes from .gw.servers where active;
+    exec serverid!attributes from .gw.servers where active,servertype=typ];
 
   if[`besteffort in key att;
-	if[-1h=type att`besteffort; besteffort:att`besteffort];
-	att:delete besteffort from att;
+  if[-1h=type att`besteffort;besteffort:att`besteffort];
+  att:delete besteffort from att;
```
Comment (ThomasSmyth) [inline code/processes/gateway.q]: 409 + 410 need additional indentation as they are only executed if 408 is true.

File: code/processes/gateway.q [hunk 3]
```diff
@@ -398,73 +402,79 @@ getserveridstype:{[att;typ]
...
   ];
   if[`attributetype in key att;
-	if[-11h=type att`attributetype; attype:att`attributetype];
-	att:delete attributetype from att;
+  if[-11h=type att`attributetype;attype:att`attributetype];
+  att:delete attributetype from att;
```
Comment (ThomasSmyth) [inline code/processes/gateway.q]: 413 + 414 need additional indentation as they are only executed if 412 is true.

File: code/processes/gateway.q [hunk 4]
```diff
@@ -398,73 +402,79 @@ getserveridstype:{[att;typ]
...
   ];
 
   res:$[attype=`independent; 
-	getserversindependent[att;servers;besteffort];
-	getserverscross[att;servers;besteffort]];
+  getserversindependent[att;servers;besteffort];
+  getserverscross[att;servers;besteffort]];
```
Comment (ThomasSmyth) [inline code/processes/gateway.q]: 418 + 419 need additional indentation as they are only executed as a result of 417.

File: code/processes/gateway.q [hunk 5]
```diff
@@ -483,13 +493,27 @@ syncexecj:{[query;servertype;joinfunction]
  $[all res[;0];
   // no errors - join the results
   [s:@[{(1b;x y)}joinfunction;res[;2];{(0b;"failed to apply supplied join function to results: ",x)}];
-   .gw.formatresponse[s 0;`sync;s 1]];
+   .gw.formatresponse[s 0;1b;s 1]];
   [failed:where not res[;0];
-   .gw.formatresponse[0b;`sync;"queries failed on server(s) ",(", " sv string exec servertype from servers where handle in handles failed),".  Error(s) were ","; " sv res[failed][;2]]]];
+   .gw.formatresponse[0b;1b;"queries failed on server(s) ",(", " sv string exec servertype from servers where handle in handles failed),".  Error(s) were ","; " sv res[failed][;2]]]
+  ]
```
Comment (ThomasSmyth) [inline code/processes/gateway.q]: Semi colon

File: code/processes/gateway.q [hunk 6]
```diff
@@ -41,13 +41,13 @@
 // each error will be prefixed with the errorprefix (currently "error: ")
 // the client should check if the result is a string, and if it is if it is prefixed with "error: "
 // errors will be returned when
```
Comment (jonnypress) [inline code/processes/gateway.q]: I think this comment section should be update to reflect the 3.6 changes - i.e. the new functionality we now have

File: code/processes/gateway.q [hunk 7]
```diff
@@ -131,7 +130,7 @@ canberun:{
  // check if it's possible to run anything
  availServers:availableserverstable[1b];
  if[0=count avail:distinct exec servertype from availServers;
-	:update required:(),available:(),handles:() from 0#0!queryqueue];
+  :update required:(),available:(),handles:() from 0#0!queryqueue];
```
Comment (ThomasSmyth) [inline code/processes/gateway.q]: New line for closing square to make it more obvious where this condition ends. I think adding an extra space indentation will make this more obvious.

File: code/processes/gateway.q [hunk 8]
```diff
@@ -222,17 +222,21 @@ sendclientreply:{[queryid;result;status]
  // if query has already been sent an error, don't send another one
  if[querydetails`error; :()];
  tosend:$[()~querydetails[`postback];
-	result;
-	(querydetails`postback),(enlist querydetails`query),enlist result];
- @[neg querydetails`clienth;.gw.formatresponse[status;`async;tosend];()]}
+  result;
+  (querydetails`postback),(enlist querydetails`query),enlist result];
```
Comment (ThomasSmyth) [inline code/processes/gateway.q]: Square brackets + extra space indentation for clarity.

File: code/processes/gateway.q [hunk 9]
```diff
@@ -222,17 +222,21 @@ sendclientreply:{[queryid;result;status]
...
+ $[querydetails`sync;
+  // return sync response
+  -30!(querydetails`clienth;not status;$[status;.gw.formatresponse[1b;1b;result];result]);
+  @[neg querydetails`clienth;.gw.formatresponse[status;0b;tosend];()]];
```
Comment (ThomasSmyth) [inline code/processes/gateway.q]: Extra space indentation here and 229.

File: code/processes/gateway.q [hunk 10]
```diff
@@ -398,73 +402,79 @@ getserveridstype:{[att;typ]
...
  [errStr:"";
  if[99h<>type servertype;
-	// its a list of servertypes e.g. `rdb`hdb
-	servertype,:();
-	missing:(servertype,()) except exec distinct servertype from .gw.servers where active;
-	if[count missing; errStr:.gw.errorprefix,"not all of the requested server types are available; missing "," " sv string missing];
-	queryattributes:()!();
+  // its a list of servertypes e.g. `rdb`hdb
+  servertype:distinct servertype,();
+  missing:(servertype,()) except exec distinct servertype from .gw.servers where active;
```
Comment (ThomasSmyth) [inline code/processes/gateway.q]: Line 440 will ensure `servertype` is a list, can trim this down to `servertype except ...`. (+ extra space indentation for clarity).

File: code/processes/gateway.q [hunk 11]
```diff
@@ -398,73 +402,79 @@ getserveridstype:{[att;typ]
...
 
 // execute a synchronous query
-syncexecj:{[query;servertype;joinfunction]
- if[not .gw.synccallsallowed;.gw.formatresponse[0b;`sync;"Synchronous calls are not allowed"]];
+syncexecjpre36:{[query;servertype;joinfunction]
+ if[not[.gw.synccallsallowed] and .z.K<3.6;.gw.formatresponse[0b;1b;"synchronous calls are not allowed"]];
  // check if the gateway allows the query to be called
- if[.gw.permissioned;if[not .pm.allowed [.z.u;query];.gw.formatresponse[0b;`sync;"User is not permissioned to run this query from the gateway"]]];
+ if[.gw.permissioned;
+  if[not .pm.allowed [.z.u;query];
```
Comment (ThomasSmyth) [inline code/processes/gateway.q]: Can clean up whitespace and given this block extra indentation.

File: code/processes/gateway.q [hunk 12]
```diff
@@ -503,8 +527,8 @@ runquery:{[]
   // get the handles to run on
   avail:availableserverstable[1b];
   IDs:$[11h=type torun`available; 
-	(exec first serverid by servertype from avail)[torun`available]; 
-	first each (exec serverid from avail) inter/: torun`available];
+  (exec first serverid by servertype from avail)[torun`available]; 
+  first each (exec serverid from avail) inter/: torun`available];
```
Comment (ThomasSmyth) [inline code/processes/gateway.q]: Extra indentation.

**Comment by jonnypress** [comment]
Ok I think this is coming together.  I was thinking about the test case you were working on earlier.  It would be nice to know if a request has come in sync or async and throw an error immediately if the wrong method is used.  I think you could modify .z.ps/.z.pg in the gateway to maintain a dictionary of callinghandle!isSync and then put a check into [a]syncexec - if it is in asyncexec and the request has come in sync, return an error synchronously.  If it has come in async and is in syncexec, return an error immediately asynchronously.  I think you'll have to wrap the definition of .z.pg/.z.ps appropriately to capture any prior definition (used for logging etc.)

File: code/processes/gateway.q (hunk 7 above)
Comment (jzabroski) [reply code/processes/gateway.q]: @ThomasSmyth Would it be possible to codify your code review guidelines as a `qlint` script?

Comment (ThomasSmyth) [reply code/processes/gateway.q]: @jzabroski development of a linter is currently under consideration.

File: code/processes/gateway.q [hunk 13]
```diff
@@ -555,9 +565,19 @@ pc:{
...
+.z.pc:{x@y;.gw.pc[y]}@[value;`.z.pc;{{[x]}}];
+.z.po:{x@y;.gw.po[y]}@[value;`.z.po;{{[x]}}];
+.z.pg:{
+  .gw.pgs[.z.w;1b];
+  :x@y;
+ }@[value;`.z.pg;{{[x]}}];
+.z.ps:{
+  .gw.pgs[.z.w;0b];
+  :x@y;
+ }@[value;`.z.ps;{{[x]}}];
```
Comment (jonnypress) [inline code/processes/gateway.q]: Ok I think the logic her is correct but could you just make the formatting consistent with the definitions of .z.po/.z.pc?   Also remove the .z.pg commented out line below

File: code/processes/gateway.q [hunk 14]
```diff
@@ -427,6 +427,11 @@ getserveridstype:{[att;typ]
 
 // execute an asynchronous query
 asyncexecjpts:{[query;servertype;joinfunction;postback;timeout;sync]
+ // Check correct function called
+ if[sync<>.gw.call .z.w;
+  @[neg .z.w;.gw.formatresponse[0b;not sync;"Incorrect function used: ",$[sync;"syncexec";"asyncexec"]];()];
```
Comment (jonnypress) [inline code/processes/gateway.q]: I think i see how this is working... if the client calls in sync, but calls asyncexec, then this: 

.gw.formatresponse[0b;not sync;"Incorrect function used: ",$[sync;"syncexec";"asyncexec"]]

will signal an error and so it won't get sent via (neg .z.w) back to the client.  Is that correct?  Could you make a comment to that effect if it is?  (it's a little tricky)

Comment (paulrdouglas) [reply code/processes/gateway.q]: Yeah that's the idea.
Yeah I'll add that now

**Pull request merged**